File: mtgSearch/src/main/java/com/dbottillo/mtgsearchfree/view/fragments/AboutFragment.java
package com.dbottillo.mtgsearchfree.view.fragments;


import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.dbottillo.mtgsearchfree.R;
import com.dbottillo.mtgsearchfree.model.storage.GeneralPreferences;
import com.dbottillo.mtgsearchfree.util.LOG;
import com.dbottillo.mtgsearchfree.util.TrackingManager;

import java.util.Calendar;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

import static android.net.Uri.parse;

public class AboutFragment extends BasicFragment implements View.OnClickListener, View.OnTouchListener {

    String[] librariesName = new String[]{"Smooth Progress Bar", "Picasso", "LeakMemory"};
    String[] librariesAuthor = new String[]{"Castorflex", "Square", "Square"};
    String[] librariesLink = new String[]{"https://github.com/castorflex/SmoothProgressBar", "http://square.github.io/picasso/", "https://github.com/square/leakcanary"};

    String versionName;

    @BindView(R.id.share_app)
    View shareApp;

    @BindView(R.id.send_feedback)
    Button sendFeedback;

    @BindView(R.id.about_version)
    TextView version;

    @BindView(R.id.copyright)
    TextView copyright;

    @BindView(R.id.libraries_container)
    LinearLayout cardContainer;

    private long firstTap;

    @Inject
    GeneralPreferences generalPreferences;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_about, container, false);

        setActionBarTitle(getString(R.string.action_about));

        ButterKnife.bind(this, v);
        version.setOnTouchListener(this);

        getMTGApp().getUiGraph().inject(this);

        versionName = "";
        try {
            versionName = getActivity().getPackageManager().getPackageInfo(getActivity().getPackageName(), 0).versionName;
            TextView version = (TextView) v.findViewById(R.id.about_version);
            version.setText(Html.fromHtml("<b>" + getString(R.string.version) + "</b>: " + versionName));
        } catch (PackageManager.NameNotFoundException e) {
            LOG.e(e);
        }

        sendFeedback.setOnClickListener(this);

        shareApp.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                TrackingManager.trackShareApp();
                Intent i = new Intent(Intent.ACTION_SEND);
                i.setType("text/plain");
                i.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.app_name));
                String url = "https://play.google.com/store/apps/details?id=com.dbottillo.mtgsearchfree";
                i.putExtra(Intent.EXTRA_TEXT, url);
                startActivity(Intent.createChooser(i, getString(R.string.share)));
            }
        });

        for (int i = 0; i < librariesName.length; i++) {
            View libraryView = View.inflate(getContext(), R.layout.row_library, null);
            TextView title = (TextView) libraryView.findViewById(R.id.library_name);
            title.setText(librariesName[i]);
            TextView author = (TextView) libraryView.findViewById(R.id.library_author);
            author.setText(librariesAuthor[i]);
            cardContainer.addView(libraryView);
            libraryView.setTag(0);
            libraryView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    int tag = (int) v.getTag();
                    Uri uri = parse(librariesLink[tag]);
                    Intent intent = new Intent(Intent.ACTION_VIEW, uri);
                    intent.addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY | Intent.FLAG_ACTIVITY_MULTIPLE_TASK);
                    startActivity(intent);
                    TrackingManager.trackAboutLibrary(librariesLink[0]);
                }
            });
        }

        copyright.setText(getString(R.string.copyright));
        return v;
    }

    @Override
    public void onClick(View v) {
        LOG.d();
        TrackingManager.trackOpenFeedback();
        Intent emailIntent = new Intent(Intent.ACTION_SENDTO, Uri.fromParts("mailto", getActivity().getString(R.string.email), null));
        String text = String.format(getString(R.string.feedback_text), versionName,
                String.valueOf(Build.VERSION.SDK_INT), Build.DEVICE, Build.MODEL, Build.PRODUCT);
        emailIntent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.feedback) + " " + getActivity().getString(R.string.app_name));
        emailIntent.putExtra(Intent.EXTRA_TEXT, Html.fromHtml(text));
        startActivity(Intent.createChooser(emailIntent, getString(R.string.send_feedback)));
    }

    @Override
    public String getPageTrack() {
        return "/about";
    }

    @Override
    public boolean onTouch(View v, MotionEvent ev) {
        switch (ev.getAction() & MotionEvent.ACTION_MASK) {
            case MotionEvent.ACTION_DOWN:
                firstTap = Calendar.getInstance().getTimeInMillis();
                break;
            case MotionEvent.ACTION_CANCEL:
            case MotionEvent.ACTION_UP:
                long diff = Calendar.getInstance().getTimeInMillis() - firstTap;
                long seconds = diff / 1000;
                if (seconds < 5) {
                    version.setOnTouchListener(null);
                    generalPreferences.setDebug();
                    Toast.makeText(getActivity(), R.string.debug_mode_active, Toast.LENGTH_LONG).show();
                }
                v.performClick();
                break;
            case MotionEvent.ACTION_MOVE:
                break;
            default:
                break;
        }
        return true;
    }
}

File: mtgSearch/src/main/kotlin/com/dbottillo/mtgsearchfree/view/fragments/AboutFragment.kt
package com.dbottillo.mtgsearchfree.view.fragments

import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.net.Uri.parse
import android.os.Build
import android.os.Bundle
import android.text.Html
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import butterknife.ButterKnife
import butterknife.OnClick
import butterknife.bindView
import com.dbottillo.mtgsearchfree.R
import com.dbottillo.mtgsearchfree.helper.LOG
import com.dbottillo.mtgsearchfree.persistence.GeneralPreferences
import com.dbottillo.mtgsearchfree.tracking.TrackingManager
import java.util.*

class AboutFragment : BasicFragment(), View.OnClickListener, View.OnTouchListener {

    internal var librariesName = arrayOf("Smooth Progress Bar", "Picasso", "LeakMemory")
    internal var librariesAuthor = arrayOf("Castorflex", "Square", "Square")
    internal var librariesLink = arrayOf("https://github.com/castorflex/SmoothProgressBar", "http://square.github.io/picasso/", "https://github.com/square/leakcanary")

    var versionName: String = ""
    private var firstTap: Long = 0

    val shareApp: View by bindView(R.id.share_app)
    val version: TextView by bindView(R.id.about_version)
    val copyright: TextView by bindView(R.id.copyright)
    val cardContainer: LinearLayout by bindView(R.id.libraries_container)

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater!!.inflate(R.layout.fragment_about, container, false)
        ButterKnife.bind(this, view)
        return view
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        setActionBarTitle(getString(R.string.action_about))

        try {
            versionName = activity.packageManager.getPackageInfo(activity.packageName, 0).versionName
            version.text = Html.fromHtml("<b>" + getString(R.string.version) + "</b>: " + versionName)
        } catch (e: PackageManager.NameNotFoundException) {
            LOG.d("[AboutFragment] exception: " + e.message);
        }

        shareApp.setOnClickListener {
            TrackingManager.trackShareApp();
            val i = Intent(Intent.ACTION_SEND)
            i.type = "text/plain"
            i.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.app_name))
            val url = "https://play.google.com/store/apps/details?id=com.dbottillo.mtgsearchfree"
            i.putExtra(Intent.EXTRA_TEXT, url)
            startActivity(Intent.createChooser(i, getString(R.string.share)))
        }

        for (i in librariesName.indices) {
            val libraryView = View.inflate(context, R.layout.row_library, null)
            val title = libraryView.findViewById(R.id.library_name) as TextView
            title.text = librariesName[i]
            val author = libraryView.findViewById(R.id.library_author) as TextView
            author.text = librariesAuthor[i]
            cardContainer.addView(libraryView)
            libraryView.tag = 0
            libraryView.setOnClickListener { v ->
                val tag = v.tag as Int
                val uri = parse(librariesLink[tag])
                val intent = Intent(Intent.ACTION_VIEW, uri)
                intent.addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY or Intent.FLAG_ACTIVITY_MULTIPLE_TASK)
                startActivity(intent)
                TrackingManager.trackAboutLibrary(librariesLink[0]);
            }
        }

        if (!GeneralPreferences.with(activity.applicationContext).isDebugEnabled) {
            version.setOnTouchListener(this)
        }

        copyright.text = getString(R.string.copyright)
    }

    @OnClick(R.id.send_feedback)
    override fun onClick(v: View) {
        TrackingManager.trackFeedback()
        val emailIntent = Intent(Intent.ACTION_SENDTO, Uri.fromParts("mailto", activity.getString(R.string.email), null))
        val text = String.format(getString(R.string.feedback_text), versionName,
                Build.VERSION.SDK_INT, Build.DEVICE, Build.MODEL, Build.PRODUCT)
        emailIntent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.feedback) + " " + activity.getString(R.string.app_name))
        emailIntent.putExtra(Intent.EXTRA_TEXT, Html.fromHtml(text))
        startActivity(Intent.createChooser(emailIntent, getString(R.string.send_feedback)))
    }

    override fun getPageTrack(): String? {
        return "/about"
    }

    override fun onTouch(v: View, ev: MotionEvent): Boolean {
        when (ev.action and MotionEvent.ACTION_MASK) {
            MotionEvent.ACTION_DOWN -> firstTap = Calendar.getInstance().timeInMillis
            MotionEvent.ACTION_CANCEL, MotionEvent.ACTION_UP -> {
                val diff = Calendar.getInstance().timeInMillis - firstTap
                val seconds = diff / 1000
                if (seconds < 5) {
                    version.setOnTouchListener(null)
                    GeneralPreferences.with(activity.applicationContext).setDebug()
                    Toast.makeText(activity, R.string.debug_mode_active, Toast.LENGTH_LONG).show()
                }
            }
            MotionEvent.ACTION_MOVE -> {
            }
            else -> {
            }
        }
        return true
    }

}
